File: golang/day14/main.go
package main

import (
	"bufio"
	"crypto/sha1"
	"fmt"
	"math"
	"os"
)

func loadData(filename string) ([]byte, error) {
	data, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer func() {
		err := data.Close()
		if err != nil {
			panic(err)
		}
	}()
	grid := make([]byte, 0)
	scanner := bufio.NewScanner(data)
	for scanner.Scan() {
		grid = append(grid, []byte(scanner.Text())...)
	}
	err = scanner.Err()
	if err != nil {
		return nil, err
	}
	return grid, nil
}

func toFlatIndex(row int, col int, height int) int {
	return row*height + col
}

func tiltNorth(grid []byte, dim int) {
	for row := 0; row < dim; row++ {
		for col := 0; col < dim; col++ {
			if grid[toFlatIndex(row, col, dim)] == byte('O') {
				currentRow := row
				for currentRow > 0 && grid[toFlatIndex(currentRow-1, col, dim)] == '.' {
					currentRow--
				}
				if currentRow != row {
					grid[toFlatIndex(currentRow, col, dim)] = byte('O')
					grid[toFlatIndex(row, col, dim)] = byte('.')
				}
			}
		}
	}
}

func calculateLoad(grid []byte, dim int) int {
	load := 0
	for row := 0; row < dim; row++ {
		for col := 0; col < dim; col++ {
			if grid[toFlatIndex(row, col, dim)] == byte('O') {
				load += dim - row
			}
		}
	}
	return load
}

func solutionPart1() {
	grid, err := loadData("input.txt")
	if err != nil {
		panic(err)
	}
	dim := int(math.Sqrt(float64(len(grid))))
	tiltNorth(grid, dim)
	load := calculateLoad(grid, dim)
	fmt.Printf("Load: %d\n", load)
}

func rotateGrid(grid []byte, dim int) {
	for row := 0; row < dim; row++ {
		for col := row + 1; col < dim; col++ {
			grid[toFlatIndex(row, col, dim)], grid[toFlatIndex(col, row, dim)] =
				grid[toFlatIndex(col, row, dim)], grid[toFlatIndex(row, col, dim)]
		}
	}
	for col := 0; col < dim/2; col++ {
		for row := 0; row < dim; row++ {
			grid[toFlatIndex(row, col, dim)], grid[toFlatIndex(row, dim-col-1, dim)] =
				grid[toFlatIndex(row, dim-col-1, dim)], grid[toFlatIndex(row, col, dim)]
		}
	}
}

func cycle(grid []byte, dim int) {
	for i := 0; i < 4; i++ {
		tiltNorth(grid, dim)
		rotateGrid(grid, dim)
	}
}

func findCycle(grid []byte, dim int) (int, int) {
	cache := make(map[[20]byte]int)
	cache[sha1.Sum(grid)] = 0
	index := 0
	for {
		cycle(grid, dim)
		if lastIndex, ok := cache[sha1.Sum(grid)]; ok {
			return index - lastIndex, index + 1
		} else {
			cache[sha1.Sum(grid)] = index
		}
		index += 1
	}
}

func solutionPart2() {
	grid, err := loadData("input.txt")
	if err != nil {
		panic(err)
	}
	dim := int(math.Sqrt(float64(len(grid))))
	remainingCycles := 1000000000
	cycleLength, currentCycle := findCycle(grid, dim)
	remainingCycles -= currentCycle
	remainingCycles -= (remainingCycles / cycleLength) * cycleLength
	for i := 0; i < remainingCycles; i++ {
		cycle(grid, dim)
	}
	load := calculateLoad(grid, dim)
	fmt.Printf("Load: %d\n", load)
}

func main() {
	solutionPart1()
	solutionPart2()
}

File: typescript/day14/main.ts
import * as fs from "fs";

function load_data(file_name: string): string[][] {
    return fs.readFileSync(file_name, 'utf8').split('\n').map(line => line.split(''));
}

function tilt_north(grid: string[][]) {
    for (let row = 0; row < grid.length; row++) {
        for (let col = 0; col < grid[row].length; col++) {
            if (grid[row][col] == "O") {
                let current_row = row
                while (current_row > 0 && grid[current_row - 1][col] == ".") {
                    current_row--;
                }
                if (current_row != row) {
                    grid[current_row][col] = "O";
                    grid[row][col] = ".";
                }
            }
        }
    }
}

function calculate_load(grid: string[][]): number {
    let load = 0;
    for (let row = 0; row < grid.length; row++) {
        for (let col = 0; col < grid[row].length; col++)
            if (grid[row][col] == "O") {
                load += (grid.length - row);
            }
    }
    return load;
}

function solution_part_1() {
    let grid = load_data("input.txt");
    tilt_north(grid);
    console.log(calculate_load(grid));
}

function rotate_clockwise(grid: string[][]) {
    for (let row = 0; row < grid.length; row++) {
        for (let col = row + 1; col < grid[row].length; col++) {
            let temp = grid[row][col];
            grid[row][col] = grid[col][row];
            grid[col][row] = temp;
        }
    }
    for (let row = 0; row < grid.length; row++) {
        grid[row].reverse();
    }
}

function cycle(grid: string[][]) {
    for (let i = 0; i < 4; i++) {
        tilt_north(grid);
        rotate_clockwise(grid);
    }
}

function find_cycle(grid: string[][]): [number, number] {
    let cache = new Map<string, number>();
    cache.set(JSON.stringify(grid), 0);
    let current_cycle = 0;
    while (true) {
        cycle(grid);
        if (cache.has(JSON.stringify(grid))) {
            return [current_cycle - cache.get(JSON.stringify(grid))!, current_cycle + 1];
        }
        cache.set(JSON.stringify(grid), current_cycle);
        current_cycle++;
    }
}

function solution_part_2() {
    let grid = load_data("input.txt");
    let [cycle_length, current_cycle] = find_cycle(grid);
    let remaining_cycles = (1000000000 - current_cycle) % cycle_length
    for (let i = 0; i < remaining_cycles; i++) {
        cycle(grid);
    }
    console.log(calculate_load(grid));
}

solution_part_1()
solution_part_2()